File: 112-path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean check(int C,TreeNode root,int T)
    {
        if(root == null)
            return false;
        C += root.val;

        if(C == T && root.left == null && root.right == null)
            return true;
        
        return check(C,root.left,T) || check(C,root.right,T);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        
        if(root == null)
            return false;
        
        int curr_sum = 0;

        return check(curr_sum,root,targetSum);
    }
}

File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool check(TreeNode *root,int curr_sum ,int targetSum)
    {
        if(!root)
            return false;
        curr_sum += root->val;
        if(curr_sum == targetSum and !root->left and !root->right)
            return true;
       
        
        

        return check(root->left,curr_sum,targetSum) or check(root->right,curr_sum,targetSum);

    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        
        if(!root)
            return false;
        
        int curr_sum = 0;

        return check(root,curr_sum,targetSum);
    }
};